Adaptive Dark colorscheme for Stmt HTML. Ability to programmatically export conceptual stmt files.

mcourteaux opened the pull request:
Colorscheme based on gruvbox-material.

[![image](https://github.com/halide/Halide/assets/845012/59956313-09e8-4039-b81f-b266c40c890b)](https://github.com/halide/Halide/assets/845012/be572828-5ddc-4f15-842d-9e0577b5029f)

Also improved assembly syntax highlighting (while reducing one HTTP fetch, as the highlight rules are now embedded in the generated file.).

steven-johnson reviewed (approved):
I'm not going to reject this, but let the record show that IMHO a light color scheme is preferable to a dark one; is there a way to have this say "prefer the system setting"?

steven-johnson commented on src/Func.h:
```diff
@@ -954,6 +954,14 @@ class Func {
+    /** Write out a conceptual representation of lowered code, before any parallel loop
+     * get factored out into separate functions, or GPU loops are offloaded to kernel code.r
+     * Useful for analyzing and debugging scheduling. Can emit html or plain text. */
+    void compile_to_conceptual_stmt(const std::string &filename,
```
This new API call should be added to the Python bindings as well

mcourteaux commented:
> I'm not going to reject this, but let the record show that IMHO a light color scheme is preferable to a dark one; is there a way to have this say "prefer the system setting"?

It follows your system preferences. It's one of those CSS media queries. You'll see in the diff.

alexreinking commented:
> It follows your system preferences. It's one of those CSS media queries. You'll see in the diff.

Can it be made into a checkbox that either enables or disables dark mode? It can use the system preference by default, but I don't want to have to change my whole system preference just to get the page to switch palates.

mcourteaux commented:
> Can it be made into a checkbox that either enables or disables dark mode?

Fair request; working on it.

mcourteaux commented:
Updated: added a toolbar with various settings. Browser persists those settings automatically (no cookie stuff). Color theme can be set to auto, or manual modes. @alexreinking @steven-johnson See video.

https://github.com/halide/Halide/assets/845012/be572828-5ddc-4f15-842d-9e0577b5029f

@steven-johnson Added those python bindings AFAICT. Not tested that tho. Just a copy-paste-and-modify.

alexreinking commented:
Looks great!

mcourteaux commented:
@steven-johnson Did this OSX buildbot fail twice on exactly the same test? This thing doesn't seem to fail on other things? It's surprising that this PR would be related but I'm not ruling anything out here. Do you have more knowledge on this failure? Please see #8332 regarding this as well.

steven-johnson commented:
> @steven-johnson Did this OSX buildbot fail twice on exactly the same test? This thing doesn't seem to fail on other things? It's surprising that this PR would be related but I'm not ruling anything out here. Do you have more knowledge on this failure? Please see #8332 regarding this as well.

Not sure, let's retry again

mcourteaux commented:
Failure does seem unrelated...

> 2024-06-28 17:09:27.935 generator_aot_gpu_multi_context_threaded[91557:5585717] Metal API Validation Enabled
> `Assertion failed: (buf2_result.all_equal(6))`, function run_kernels_on_thread, file gpu_multi_context_threaded_aottest.cpp, line 237.
> 473/474 Test #631: generator_aotcpp_gpu_multi_context_threaded ................Subprocess aborted***Exception:  32.16 sec
> 2024-06-28 17:09:27.944 generator_aotcpp_gpu_multi_context_threaded[91558:5585726] Metal API Validation Enabled
> `Assertion failed: (buf1_result.all_equal(val + 2))`, function run_kernels_on_thread, file gpu_multi_context_threaded_aottest.cpp, line 236.

Can this be merged, or should something change?
